common/configloader/configloader.go

```diff
@@ -0,0 +1,173 @@
+package configloader
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+)
+
+var ErrNotSet = errors.New("The value for the given key was not set")
+var ErrDecodeFailed = errors.New("The value for the given key could not be decoded")
+var ErrLoadFailed = errors.New("Unable to load config")
+
+/*
+	Used to load a key value configuration
+*/
+type ConfigLoader struct {
+	configPath   string
+	parsedConfig map[string]*json.RawMessage
+}
+
+/*
+	Loads the configuration at the given path into a ConfigLoader struct
+	Supported uri schemes are: s3, file, https
+*/
+func Load(configPath string) (*ConfigLoader, error) {
+	uri, err := url.Parse(configPath)
+
+	if err != nil {
+		return nil, ErrLoadFailed
+	}
+
+	var configContents []byte
+
+	switch uri.Scheme {
+	case "s3":
+		configContents, err = loadFromS3(configPath)
+	case "file":
+		configContents, err = loadFromFile(configPath)
+	case "https":
+		configContents, err = loadFromHttps(configPath)
+	default:
+		return nil, ErrLoadFailed
+	}
+
+	if err != nil {
+		return nil, ErrLoadFailed
+	}
+
+	loader := ConfigLoader{
+		configPath: configPath,
+	}
+
+	err = json.Unmarshal(configContents, &loader.parsedConfig)
+
+	if err != nil {
+		return nil, ErrLoadFailed
+	}
+
+	return &loader, nil
+}
+
+/*
+	Returns the value associated with a given key as a string
+	Environment variables will override configuration
+*/
+func (loader *ConfigLoader) Get(key string) (string, error) {
+	value, exists := os.LookupEnv(key)
+
+	if exists {
+		return value, nil
+	}
+
+	rawValue, exists := loader.parsedConfig[key]
+
+	if !exists {
+		return "", ErrNotSet
+	}
+
+	err := json.Unmarshal(*rawValue, &value)
+
+	if err != nil {
+		return "", ErrDecodeFailed
+	}
+
+	return value, nil
+}
+
+/*
+	Parses the value the given key into the given interface{}
+	Environment variables will override configutation
+*/
+func (loader *ConfigLoader) ParseInto(key string, out interface{}) error {
```

Comment:
What is the purpose of this method?

Reply:
If we have json more complex than a key-value pair, then it is impossible to parse until we have a struct definition to parse into. So we when loading the json we parse the top level and and further nesting can be parsed with this method. For an example of it's usage you can see the associated test.